src/App.tsx

```diff
@@ -1,24 +1,146 @@
-import React from 'react';
-import logo from './logo.svg';
+import React, { useState, useEffect, useMemo } from 'react';
 import './App.css';
+import Navigation from "./components/Navigation";
+import Scores from "./components/Scores";
+import Home from './components/Home';
+import Total from "./components/Total";
+import {BrowserRouter as Router, Switch, Route} from "react-router-dom";
+import {IPlayer, IScore, IRoutes} from "./types";
+import ROUTES from './routes';
+import {nanoid} from 'nanoid';
+import {debounce} from 'debounce';
+
+const scoreTemplate = {
+	military: 0,
+	treasury: 0,
+	wonders: 0,
+	civilian: 0,
+	commerce: 0,
+	guild: 0,
+	compass: 0,
+	tablet: 0,
+	gear: 0
+};
 
 function App() {
+	const [players, setPlayers] = useState<IPlayer[]>([]);
+
+	useEffect(() => {
+		saveToStorage(players);
+	});
+
+	useEffect(() => {
+		restoreFromStorage();
+	}, []);
+
+	const saveToStorage = useMemo(() => debounce((players: IPlayer[]) => {
+		localStorage.setItem('players', JSON.stringify(players));
+	}, 500), []);
+
+	function restoreFromStorage(): void {
+		const players = localStorage.getItem('players');
+
+		if (players) {
+			setPlayers(JSON.parse(players));
+		}
+	}
+
+	function addPlayer(name: string) {
+		if (name) {
+			setPlayers([
+				...players, {
+					id: nanoid(),
+					name,
+					score: {
+						...scoreTemplate
+					}
+				}
+			]);
+		}
+	}
+
+	function deletePlayer(id: string): void {
+		if (id) {
+			setPlayers([
+				...players.filter(player => {
+					return player.id !== id
+				})
+			]);
+		}
+	}
+
+	function updatePlayerScore(id: string, payload: IPlayer) {
+		if (id) {
+			setPlayers([
+				...players.map(player => {
+					if (player.id === id) {
+						return payload;
+					} else {
+						return player;
+					}
+				}),
+			]);
+		}
+	}
+
+	function onChange(id: string, scoreKey: keyof IScore, value: number) {
+		const player = players.find(player => player.id === id);
+
+		if (player) {
+			updatePlayerScore(id, {
+				...player,
+				score: {
+					...player.score,
+					[scoreKey]: value
+				}
+			});
+		}
+	}
```

Comment:
You do not need to make two scans of the 'players' array in two methods above.
First, find the *index* of the player by its id.
Next, create new player from the players[i].
Finally, splice the players array with the new value in order to make react find the object has changed.

Reply:
Fixed